hail/src/main/scala/is/hail/expr/ir/lowering/CanLowerEfficiently.scala

```diff
@@ -0,0 +1,111 @@
+package is.hail.expr.ir.lowering
+
+import is.hail.HailContext
+import is.hail.expr.ir._
+import is.hail.expr.ir.functions.TableToValueFunction
+import is.hail.io.TextMatrixReader
+import is.hail.io.bgen.MatrixBGENReader
+import is.hail.io.plink.MatrixPLINKReader
+import is.hail.io.vcf.MatrixVCFReader
+import is.hail.methods.{ForceCountTable, NPartitionsTable}
+
+object CanLowerEfficiently {
+  def apply(ir0: BaseIR): Option[String] = {
+
+    if (HailContext.getFlag("no_whole_stage_codegen") != null)
+      return Some("flag 'no_whole_stage_codegen' is enabled")
+
+    var prohibitiveReason: Option[String] = None
+
+    def fail(reason: String): Unit = {
+      assert(prohibitiveReason.isEmpty)
+      prohibitiveReason = Some(reason)
+    }
+
+    def recur(ir: BaseIR): Unit = {
+
+      ir match {
+        case TableRead(_, _, _: TableNativeReader) =>
+        case TableRead(_, _, _: TableNativeZippedReader) =>
+        case TableRead(_, _, _: TextTableReader) =>
+        case TableRead(_, _, _: TextMatrixReader) =>
+        case TableRead(_, _, _: MatrixPLINKReader) =>
+        case TableRead(_, _, _: MatrixVCFReader) =>
+        case TableRead(_, _, _: MatrixBGENReader) =>
+          fail(s"no lowering for MatrixBGENReader")
+        case TableRead(_, _, _: TableFromBlockMatrixNativeReader) =>
+          fail(s"no lowering for TableFromBlockMatrixNativeReader")
+
+        case t: TableLiteral =>
+        case t: TableRepartition => fail(s"TableRepartition has no lowered implementation")
+        case t: TableParallelize =>
+        case t: TableRange =>
+        case TableKeyBy(child, keys, isSorted) =>
+          val doesStrangeDNDArrayOneToOneRekey = isSorted &&
+            child.typ.key.zip(keys).takeWhile { case (l, r) => l == r }.isEmpty
+          if (doesStrangeDNDArrayOneToOneRekey)
+            fail(s"TableKeyBy cannot lower the one-to-one rekey done in dndarrays")
+        case t: TableOrderBy =>
+        case t: TableFilter =>
+        case t: TableHead =>
+        case t: TableTail => fail("TableTail has no short-circuit using known partition counts")
+        case t: TableJoin =>
+        case t: TableIntervalJoin => fail(s"TableIntervalJoin has no lowered implementation")
+        case t: TableMultiWayZipJoin =>
+        case t: TableLeftJoinRightDistinct =>
+        case t: TableMapPartitions =>
+        case t: TableMapRows => if (ContainsScan(t.newRow)) fail("TableMapRows does not have a scalable implementation of scans")
+        case t: TableMapGlobals =>
+        case t: TableExplode =>
+        case t: TableUnion if t.children.length > 16 => fail(s"TableUnion lowering generates deeply nested IR if it has many children")
```

Comment:
Could you explain where the "deeply nested IR" comes into play? Lookinag the at the lowering rule for `TableUnion` is isn't apparent to me.

Reply:
by "nesting depth" I'm referring to the longest path between the root of the IR tree and a leaf. Since most of our compiler passes are written as recursive, it's possible/easy for us to blow stack size limits when generating deep IRs.

TableUnion generates IR with depth proportional to the number of `t.children`, due to the way the Let bindings get placed. This isn't an inherent problem, but interacts badly with the limitations of our compiler. We should continue rewriting passes to work iteratively using a stack on the JVM heap rather than using the JVM function stack directly.

Reply:
So the lowering of `TableUnion` doesn't need to change, just our individual lowering passes?

We could also introduce a `Let*` node for nodes that introduce many bindings at once to prevent this. Probably trickier to optimize around though.

Reply:
That's right, the lowering is correct, but we need to support deeply nested IRs in other compiler infrastructure.

Yeah, tougher to optimize with that node. At some point soon, we're going to discuss the Future of IR™ with Cotton's proposal for a graph-based IR which we will probably introduce to Scala from the bottom up.